File: frontend/src/utils/indicators.ts
// 纯技术指标计算：输入按时间升序排列的收盘价（或完整 K 线），返回与输入等长
// 的数值数组，头部预热期（数据不足以算出第一个值的那些位置）填 null，交由
// 调用方（ChartsPage）过滤/转换成 lightweight-charts 的 series 数据点。
// 这里只做纯数学，不碰任何图表 API，方便单独验证正确性。
//
// Pure technical-indicator math: takes an ascending-by-time series of closes
// (or full candles), returns an array of the same length with `null` for the
// warm-up head (not enough data yet for a value). The caller (ChartsPage)
// filters/converts these into lightweight-charts series data points. Kept
// free of any charting API so the math itself is easy to verify in isolation.
import type { Candle } from '../api/types'

// 简单移动平均 / simple moving average
export function sma(values: number[], period: number): (number | null)[] {
  const out: (number | null)[] = new Array(values.length).fill(null)
  let sum = 0
  for (let i = 0; i < values.length; i++) {
    sum += values[i]
    if (i >= period) sum -= values[i - period]
    if (i >= period - 1) out[i] = sum / period
  }
  return out
}

// 指数移动平均：首个有效值用同周期 SMA 作种子（业界通用做法），此后递推。
// exponential moving average: seeded with the SMA of the first `period`
// values (the standard convention), recursive thereafter.
export function ema(values: number[], period: number): (number | null)[] {
  const out: (number | null)[] = new Array(values.length).fill(null)
  if (values.length < period) return out
  const k = 2 / (period + 1)
  let seed = 0
  for (let i = 0; i < period; i++) seed += values[i]
  seed /= period
  out[period - 1] = seed
  let prev = seed
  for (let i = period; i < values.length; i++) {
    prev = values[i] * k + prev * (1 - k)
    out[i] = prev
  }
  return out
}

function stddevAroundSma(values: number[], smaArr: (number | null)[], period: number): (number | null)[] {
  const out: (number | null)[] = new Array(values.length).fill(null)
  for (let i = period - 1; i < values.length; i++) {
    const mean = smaArr[i]
    if (mean == null) continue
    let sq = 0
    for (let j = i - period + 1; j <= i; j++) sq += (values[j] - mean) ** 2
    out[i] = Math.sqrt(sq / period)
  }
  return out
}

export interface BollBands {
  mid: (number | null)[]
  upper: (number | null)[]
  lower: (number | null)[]
}

// 布林带：中轨=SMA，上下轨=中轨 ± mult 倍标准差 / Bollinger Bands: mid=SMA,
// upper/lower = mid +/- mult standard deviations
export function bollinger(values: number[], period = 20, mult = 2): BollBands {
  const mid = sma(values, period)
  const sd = stddevAroundSma(values, mid, period)
  const upper = mid.map((m, i) => (m == null || sd[i] == null ? null : m + mult * (sd[i] as number)))
  const lower = mid.map((m, i) => (m == null || sd[i] == null ? null : m - mult * (sd[i] as number)))
  return { mid, upper, lower }
}

// RSI（Wilder 平滑，与主流平台一致，不是简单移动平均版本）
// RSI using Wilder's smoothing (matches mainstream platforms; not the plain
// moving-average variant)
export function rsi(values: number[], period = 14): (number | null)[] {
  const out: (number | null)[] = new Array(values.length).fill(null)
  if (values.length <= period) return out
  let gainSum = 0
  let lossSum = 0
  for (let i = 1; i <= period; i++) {
    const diff = values[i] - values[i - 1]
    if (diff >= 0) gainSum += diff
    else lossSum -= diff
  }
  let avgGain = gainSum / period
  let avgLoss = lossSum / period
  out[period] = avgLoss === 0 ? 100 : 100 - 100 / (1 + avgGain / avgLoss)
  for (let i = period + 1; i < values.length; i++) {
    const diff = values[i] - values[i - 1]
    const gain = diff > 0 ? diff : 0
    const loss = diff < 0 ? -diff : 0
    avgGain = (avgGain * (period - 1) + gain) / period
    avgLoss = (avgLoss * (period - 1) + loss) / period
    out[i] = avgLoss === 0 ? 100 : 100 - 100 / (1 + avgGain / avgLoss)
  }
  return out
}

export interface MacdResult {
  macd: (number | null)[]
  signal: (number | null)[]
  hist: (number | null)[]
}

// MACD：DIF=快慢 EMA 之差，DEA(signal)=DIF 的 EMA，柱=DIF-DEA
// MACD: DIF = fast EMA - slow EMA, signal (DEA) = EMA of DIF, histogram = DIF - signal
export function macd(values: number[], fast = 12, slow = 26, signalPeriod = 9): MacdResult {
  const fastEma = ema(values, fast)
  const slowEma = ema(values, slow)
  const macdLine: (number | null)[] = values.map((_, i) =>
    fastEma[i] == null || slowEma[i] == null ? null : (fastEma[i] as number) - (slowEma[i] as number)
  )
  // signal 是对 macdLine 的 EMA，但 macdLine 头部有 null（慢线预热期），要先
  // 抽出从第一个非空值开始的连续段再算 EMA，再把结果按原位置拼回去。
  // signal is an EMA over macdLine, but macdLine has a null head (the slow
  // EMA's warm-up); pull out the dense run starting at the first non-null
  // value, run EMA over that, then splice the result back into position.
  const firstValid = macdLine.findIndex((v) => v != null)
  const signal: (number | null)[] = new Array(values.length).fill(null)
  if (firstValid >= 0) {
    const dense = macdLine.slice(firstValid) as number[]
    const denseSignal = ema(dense, signalPeriod)
    for (let i = 0; i < denseSignal.length; i++) signal[firstValid + i] = denseSignal[i]
  }
  const hist: (number | null)[] = values.map((_, i) =>
    macdLine[i] == null || signal[i] == null ? null : (macdLine[i] as number) - (signal[i] as number)
  )
  return { macd: macdLine, signal, hist }
}

// 唐奇安通道上轨：每个位置取该 bar 之前（不含当根）最近 period 根的最高价。
// 排除当根与后端 indicators.py 的 donchian_high 一致，必须一致——回测的突破条件
// 判断的是"当根价格是否超过之前的极值"，把当根算进去突破永远不成立，画出来的线
// 就会比判定所用的那条高，看上去像"标记打在没突破的地方"。
// Donchian upper band: for each position, the highest high of the `period` bars
// strictly before it. Excluding the current bar matches the backend's
// donchian_high and has to: the breakout condition asks whether the current price
// exceeds the prior extreme, and including the current bar makes a breakout
// impossible — the drawn line would sit above the one actually used for the
// verdict, making markers look like they fired without a breakout.
export function donchianHigh(highs: number[], period: number): (number | null)[] {
  const out: (number | null)[] = new Array(highs.length).fill(null)
  for (let i = period; i < highs.length; i++) {
    let hi = -Infinity
    for (let j = i - period; j < i; j++) if (highs[j] > hi) hi = highs[j]
    out[i] = hi
  }
  return out
}

// 唐奇安通道下轨，见 donchianHigh / Donchian lower band; see donchianHigh
export function donchianLow(lows: number[], period: number): (number | null)[] {
  const out: (number | null)[] = new Array(lows.length).fill(null)
  for (let i = period; i < lows.length; i++) {
    let lo = Infinity
    for (let j = i - period; j < i; j++) if (lows[j] < lo) lo = lows[j]
    out[i] = lo
  }
  return out
}

// 从完整 K 线数组里取收盘价，供上面几个函数直接使用 / pull closes out of full
// candles for the functions above to consume directly
export function closes(bars: Candle[]): number[] {
  return bars.map((b) => b.c)
}

// ───────── 2026-09-08 指标库扩充 / indicator library expansion ─────────
// 下面这些都吃完整 K 线（要用到高低价 / 成交量），返回与输入等长、预热期为 null 的数组。
// These take full candles (they need highs/lows/volume) and return same-length
// arrays with null over the warm-up head, like everything above.

// 真实波幅（Wilder 平滑）/ Average True Range with Wilder smoothing
export function atr(bars: Candle[], period = 14): (number | null)[] {
  const out: (number | null)[] = new Array(bars.length).fill(null)
  if (bars.length <= period) return out
  const tr = bars.map((b, i) => (i === 0 ? b.h - b.l : Math.max(b.h - b.l, Math.abs(b.h - bars[i - 1].c), Math.abs(b.l - bars[i - 1].c))))
  let sum = 0
  for (let i = 1; i <= period; i++) sum += tr[i]
  let prev = sum / period
  out[period] = prev
  for (let i = period + 1; i < bars.length; i++) {
    prev = (prev * (period - 1) + tr[i]) / period
    out[i] = prev
  }
  return out
}

export interface SuperTrendResult {
  // 多头段（线在价格下方）与空头段（线在价格上方）分成两条，各自在不属于自己的
  // 区间填 null，这样两条不同颜色的 series 拼起来就是一条会变色的线。
  // Bull segments (line below price) and bear segments (above) as two arrays,
  // each null where the other owns the bar, so two colored series read as one
  // line that changes color.
  bull: (number | null)[]
  bear: (number | null)[]
}

// 超级趋势：ATR 通道 + 方向翻转规则（标准实现）/ SuperTrend: ATR bands with the standard flip rule
export function superTrend(bars: Candle[], period = 10, mult = 3): SuperTrendResult {
  const n = bars.length
  const bull: (number | null)[] = new Array(n).fill(null)
  const bear: (number | null)[] = new Array(n).fill(null)
  const a = atr(bars, period)
  let finalUpper = 0
  let finalLower = 0
  let dir = 1
  let started = false
  for (let i = 0; i < n; i++) {
    const av = a[i]
    if (av == null) continue
    const hl2 = (bars[i].h + bars[i].l) / 2
    const upper = hl2 + mult * av
    const lower = hl2 - mult * av
    if (!started) {
      finalUpper = upper
      finalLower = lower
      started = true
    } else {
      const prevClose = bars[i - 1].c
      finalUpper = upper < finalUpper || prevClose > finalUpper ? upper : finalUpper
      finalLower = lower > finalLower || prevClose < finalLower ? lower : finalLower
    }
    const c = bars[i].c
    if (dir === 1 && c < finalLower) dir = -1
    else if (dir === -1 && c > finalUpper) dir = 1
    if (dir === 1) bull[i] = finalLower
    else bear[i] = finalUpper
  }
  return { bull, bear }
}

// 抛物线转向 SAR（Wilder）/ Parabolic SAR
export function parabolicSar(bars: Candle[], step = 0.02, max = 0.2): (number | null)[] {
  const n = bars.length
  const out: (number | null)[] = new Array(n).fill(null)
  if (n < 2) return out
  let up = bars[1].c >= bars[0].c
  let sar = up ? bars[0].l : bars[0].h
  let ep = up ? bars[0].h : bars[0].l
  let af = step
  for (let i = 1; i < n; i++) {
    const b = bars[i]
    sar = sar + af * (ep - sar)
    if (up) {
      sar = Math.min(sar, bars[i - 1].l, i >= 2 ? bars[i - 2].l : bars[i - 1].l)
      if (b.l < sar) {
        up = false
        sar = ep
        ep = b.l
        af = step
      } else if (b.h > ep) {
        ep = b.h
        af = Math.min(max, af + step)
      }
    } else {
      sar = Math.max(sar, bars[i - 1].h, i >= 2 ? bars[i - 2].h : bars[i - 1].h)
      if (b.h > sar) {
        up = true
        sar = ep
        ep = b.h
        af = step
      } else if (b.l < ep) {
        ep = b.l
        af = Math.min(max, af + step)
      }
    }
    out[i] = sar
  }
  return out
}

// 成交量加权均价，按自然日（UTC+8，与图表坐标轴同一时区）重置锚点。
// VWAP anchored to the natural day (UTC+8, same zone as the chart axis).
export function vwap(bars: Candle[]): (number | null)[] {
  const out: (number | null)[] = new Array(bars.length).fill(null)
  const TZ = 8 * 3600
  let day = -1
  let pv = 0
  let vol = 0
  for (let i = 0; i < bars.length; i++) {
    const b = bars[i]
    const d = Math.floor((b.t + TZ) / 86400)
    if (d !== day) {
      day = d
      pv = 0
      vol = 0
    }
    const tp = (b.h + b.l + b.c) / 3
    pv += tp * b.v
    vol += b.v
    out[i] = vol > 0 ? pv / vol : null
  }
  return out
}

export interface KdjResult {
  k: (number | null)[]
  d: (number | null)[]
  j: (number | null)[]
}

// KDJ：RSV 的两级平滑（国内软件通用的 9/3/3 口径），J = 3K − 2D。
// KDJ: two-stage smoothed RSV (the common 9/3/3 convention), J = 3K − 2D.
export function kdj(bars: Candle[], period = 9, kSmooth = 3, dSmooth = 3): KdjResult {
  const n = bars.length
  const k: (number | null)[] = new Array(n).fill(null)
  const d: (number | null)[] = new Array(n).fill(null)
  const j: (number | null)[] = new Array(n).fill(null)
  let kPrev = 50
  let dPrev = 50
  for (let i = period - 1; i < n; i++) {
    let hi = -Infinity
    let lo = Infinity
    for (let x = i - period + 1; x <= i; x++) {
      if (bars[x].h > hi) hi = bars[x].h
      if (bars[x].l < lo) lo = bars[x].l
    }
    const rsv = hi === lo ? 50 : ((bars[i].c - lo) / (hi - lo)) * 100
    kPrev = ((kSmooth - 1) * kPrev + rsv) / kSmooth
    dPrev = ((dSmooth - 1) * dPrev + kPrev) / dSmooth
    k[i] = kPrev
    d[i] = dPrev
    j[i] = 3 * kPrev - 2 * dPrev
  }
  return { k, d, j }
}

// 顺势指标 CCI = (TP − SMA(TP)) / (0.015 × 平均绝对偏差)
// Commodity Channel Index = (TP − SMA(TP)) / (0.015 × mean absolute deviation)
export function cci(bars: Candle[], period = 20): (number | null)[] {
  const tp = bars.map((b) => (b.h + b.l + b.c) / 3)
  const mean = sma(tp, period)
  return tp.map((v, i) => {
    const m = mean[i]
    if (m == null) return null
    let dev = 0
    for (let x = i - period + 1; x <= i; x++) dev += Math.abs(tp[x] - m)
    dev /= period
    return dev === 0 ? 0 : (v - m) / (0.015 * dev)
  })
}

// 威廉指标 %R = (Hn − C) / (Hn − Ln) × −100，范围 −100 到 0
// Williams %R = (Hn − C) / (Hn − Ln) × −100, ranging −100 to 0
export function williamsR(bars: Candle[], period = 14): (number | null)[] {
  return bars.map((b, i) => {
    if (i < period - 1) return null
    let hi = -Infinity
    let lo = Infinity
    for (let x = i - period + 1; x <= i; x++) {
      if (bars[x].h > hi) hi = bars[x].h
      if (bars[x].l < lo) lo = bars[x].l
    }
    return hi === lo ? -50 : ((hi - b.c) / (hi - lo)) * -100
  })
}

// 能量潮：收涨累加成交量，收跌累减 / On-Balance Volume
export function obv(bars: Candle[]): (number | null)[] {
  const out: (number | null)[] = new Array(bars.length).fill(null)
  let acc = 0
  for (let i = 0; i < bars.length; i++) {
    if (i > 0) {
      if (bars[i].c > bars[i - 1].c) acc += bars[i].v
      else if (bars[i].c < bars[i - 1].c) acc -= bars[i].v
    }
    out[i] = acc
  }
  return out
}

File: backend/app/services/strategy/indicators.py
"""指标数学：从 strategy_engine.py 抽出的纯函数，输入输出都是等长序列。
Indicator math: pure functions lifted out of strategy_engine.py; inputs and
outputs are equal-length series."""
import math


def sma(values: list[float], period: int) -> list[float | None]:
    out: list[float | None] = [None] * len(values)
    total = 0.0
    for i, v in enumerate(values):
        total += v
        if i >= period:
            total -= values[i - period]
        if i >= period - 1:
            out[i] = total / period
    return out


def ema(values: list[float], period: int) -> list[float | None]:
    out: list[float | None] = [None] * len(values)
    if len(values) < period:
        return out
    seed = sum(values[:period]) / period
    out[period - 1] = seed
    k = 2 / (period + 1)
    prev = seed
    for i in range(period, len(values)):
        prev = values[i] * k + prev * (1 - k)
        out[i] = prev
    return out


def rsi(values: list[float], period: int) -> list[float | None]:
    out: list[float | None] = [None] * len(values)
    if len(values) <= period:
        return out
    gain_sum = loss_sum = 0.0
    for i in range(1, period + 1):
        diff = values[i] - values[i - 1]
        if diff >= 0:
            gain_sum += diff
        else:
            loss_sum -= diff
    avg_gain, avg_loss = gain_sum / period, loss_sum / period
    out[period] = 100.0 if avg_loss == 0 else 100 - 100 / (1 + avg_gain / avg_loss)
    for i in range(period + 1, len(values)):
        diff = values[i] - values[i - 1]
        gain = diff if diff > 0 else 0.0
        loss = -diff if diff < 0 else 0.0
        avg_gain = (avg_gain * (period - 1) + gain) / period
        avg_loss = (avg_loss * (period - 1) + loss) / period
        out[i] = 100.0 if avg_loss == 0 else 100 - 100 / (1 + avg_gain / avg_loss)
    return out


def bollinger(
    values: list[float], period: int, mult: float
) -> tuple[list[float | None], list[float | None], list[float | None]]:
    """布林带：返回 (上轨, 中轨, 下轨)。中轨即 SMA，用的是总体标准差(分母 n)，
    与前端 indicators.ts 口径一致。
    Bollinger bands: returns (upper, middle, lower). The middle band is the SMA;
    the deviation is the population stdev (divisor n), matching the frontend's
    indicators.ts.

    标准差用滑动平方和递推,而不是每根 bar 调一次 statistics.pstdev:后者每根都要
    拷一份长度 period 的切片并走高精度求和,整体 O(n×period),实测 5 万根时占到
    单条策略求值的 4.98 秒(其余五类指标合计仅 0.78 秒)。递推是 O(n),约 20 毫秒。

    递推里所有累加量都先减去一个固定偏移 origin(取序列首个值)再累加。这一步不是
    可选的微调:方差公式 E[x²]-E[x]² 要把两个量级相近的大数相减,当窗口内波动远
    小于价格本身时(金价 2000 上下、20 根内只走几毛钱,正是最常见的情形),有效位
    会被大量抵消。实测拿 Fraction 精确算术当基准:不减偏移的最大相对误差 6.7e-08,
    减了偏移是 2.6e-14,差 2600 倍,而耗时只从 18 毫秒变成 20 毫秒。

    也试过滑动窗口 Welford(9.5e-12)——反而不如减偏移,因为它移除旧值时要做逆向
    更新,那一步自身会累积误差,不像只增不减的 Welford 那样稳。逐窗口两趟纯浮点
    能到 2.1e-16(浮点极限),但那是 O(n×period),period=100 时要 253 毫秒。

    方差仍做 max(0.0, ...) 钳位:平盘序列理论方差为 0,浮点下可能落在 -1e-13 这类
    极小负数上,不钳位 sqrt 会抛 ValueError。那不是防御性样板代码。

    The stdev uses a rolling sum-of-squares recurrence rather than calling
    statistics.pstdev per bar: the latter copies a period-length slice and runs a
    high-precision summation for every bar, making it O(n*period) — measured at
    4.98s of a single strategy's evaluation over 50k bars, against 0.78s for the
    other five indicator families combined. The recurrence is O(n), around 20ms.

    Every accumulator subtracts a fixed `origin` (the first value) before summing.
    This is not an optional tweak: the variance form E[x^2]-E[x]^2 subtracts two
    similarly-sized large numbers, and when in-window movement is far smaller than
    the price level (gold near 2000 moving cents over 20 bars — the common case),
    significant digits cancel away. Measured against exact Fraction arithmetic: max
    relative error is 6.7e-08 without the offset and 2.6e-14 with it, a 2600x
    difference, for 18ms to 20ms.

    A sliding-window Welford was also tried (9.5e-12) and is *worse* than the
    offset: removing the outgoing value needs a reverse update that accumulates
    error of its own, unlike increment-only Welford. A per-window two-pass float
    sum reaches 2.1e-16 (the floating-point floor) but is O(n*period), costing
    253ms at period=100.

    The variance keeps a max(0.0, ...) clamp: a flat series has zero variance in
    theory and can land on values like -1e-13 in floating point, where sqrt would
    raise ValueError. That clamp is not defensive boilerplate.
    """
    n = len(values)
    middle = sma(values, period)
    upper: list[float | None] = [None] * n
    lower: list[float | None] = [None] * n
    if period <= 0 or n < period:
        return upper, middle, lower

    # 偏移基准。取首个值即可:同一序列内的价格不会跨数量级,减掉它就足以把累加量
    # 压到与窗口波动同量级。理由见上面 docstring。
    # Offset origin. The first value suffices: prices within one series don't span
    # orders of magnitude, so subtracting it brings the accumulators down to the
    # same scale as the in-window movement. Rationale in the docstring.
    origin = values[0]
    total = 0.0
    total_sq = 0.0
    for i, raw in enumerate(values):
        v = raw - origin
        total += v
        total_sq += v * v
        if i >= period:
            dropped = values[i - period] - origin
            total -= dropped
            total_sq -= dropped * dropped
        if i < period - 1:
            continue
        m = middle[i]
        if m is None:
            continue
        # 方差对平移不变,所以偏移后的累加量直接给出原序列的方差。
        # Variance is translation-invariant, so the offset accumulators give the
        # original series' variance directly.
        variance = max(0.0, total_sq / period - (total / period) ** 2)
        sd = math.sqrt(variance)
        upper[i] = m + mult * sd
        lower[i] = m - mult * sd
    return upper, middle, lower


def macd(
    values: list[float], fast: int, slow: int, signal_period: int
) -> tuple[list[float | None], list[float | None]]:
    """MACD：DIF=快慢 EMA 之差，DEA(signal)=DIF 的 EMA——移植自 indicators.ts 的
    macd()，同样先抽出 DIF 里第一个非空值开始的连续段再算 EMA，再按原位置拼回去
    （慢线预热期会让 DIF 数组开头有一段 None）。
    MACD: DIF = fast EMA - slow EMA, signal (DEA) = EMA of DIF — ported from
    indicators.ts's macd(), pulling out the dense run starting at DIF's first
    non-null value before running EMA over it (the slow EMA's warm-up leaves a
    None head), then splicing the result back into position."""
    fast_ema = ema(values, fast)
    slow_ema = ema(values, slow)
    macd_line: list[float | None] = [
        None if fast_ema[i] is None or slow_ema[i] is None else fast_ema[i] - slow_ema[i]
        for i in range(len(values))
    ]
    first_valid = next((i for i, v in enumerate(macd_line) if v is not None), -1)
    signal: list[float | None] = [None] * len(values)
    if first_valid >= 0:
        dense = macd_line[first_valid:]
        dense_signal = ema(dense, signal_period)
        for i, v in enumerate(dense_signal):
            signal[first_valid + i] = v
    return macd_line, signal


def atr(
    highs: list[float], lows: list[float], closes: list[float], period: int
) -> list[float | None]:
    """ATR：真实波幅的 Wilder 平滑均值，与 rsi() 同一套平滑口径。

    首根 bar 没有前收盘价，真实波幅退化为当根高低差。前 period 根用简单平均
    做种子，其后按 Wilder 递推。

    ATR: Wilder-smoothed average true range, same smoothing convention as
    rsi(). The first bar has no previous close, so its true range degenerates
    to the high-low span. The first `period` bars seed with a simple average;
    subsequent bars use Wilder's recursion.
    """
    n = len(closes)
    out: list[float | None] = [None] * n
    if n < period or period < 1:
        return out
    trs: list[float] = []
    for i in range(n):
        if i == 0:
            trs.append(highs[i] - lows[i])
        else:
            prev_close = closes[i - 1]
            trs.append(max(highs[i] - lows[i], abs(highs[i] - prev_close), abs(lows[i] - prev_close)))
    prev = sum(trs[:period]) / period
    out[period - 1] = prev
    for i in range(period, n):
        prev = (prev * (period - 1) + trs[i]) / period
        out[i] = prev
    return out


def donchian_high(highs: list[float], period: int) -> list[float | None]:
    """每个位置取该 bar 之前(不含当根)最近 period 根的最高价。

    排除当根是刻意的：突破类条件要判断"当根价格是否超过之前的极值"，把当根
    自己算进极值会让突破永远不成立。等价于原 strategy_engine._rolling_max_excl
    的逐 bar 版本，这里一次算出整条序列供规则引擎消费。

    For each position, the highest high of the `period` bars strictly before
    it. Excluding the current bar is deliberate: a breakout condition asks
    whether the current price exceeds the prior extreme, and including the
    current bar in that extreme would make a breakout impossible. Equivalent
    to the per-bar _rolling_max_excl in the old strategy_engine, computed as a
    whole series here for the rule engine to consume.
    """
    out: list[float | None] = [None] * len(highs)
    for i in range(period, len(highs)):
        out[i] = max(highs[i - period : i])
    return out


def donchian_low(lows: list[float], period: int) -> list[float | None]:
    """每个位置取该 bar 之前(不含当根)最近 period 根的最低价。见 donchian_high。
    Lowest low of the `period` bars strictly before each position. See
    donchian_high."""
    out: list[float | None] = [None] * len(lows)
    for i in range(period, len(lows)):
        out[i] = min(lows[i - period : i])
    return out


def cmp_with_tol(a: float, b: float, rel_tol: float = 1e-9) -> int:
    """带容差比较两个浮点数,返回 -1/0/1。

    两条完全由同一批重复价格递推出来的 EMA,理论上应该分毫不差——但
    `prev = value*k + prev*(1-k)` 这种递归乘加运算,只要平滑系数 k=2/(period+1)
    在二进制浮点下不能精确表示(取决于具体的 period,不可预测哪些值会中招,
    实测 period=30 就会、12/26/9/14/50 不会),就会残留 1e-13~1e-14 级别的
    误差(比如 100.0 变成 100.00000000000001)。这点误差如果直接拿去和另一
    条均线做严格的 `<`/`>` 比较,会被误判成"刚刚穿越",凭空报出一个不存在
    的交叉。容差取两数量级的 1e-9 倍(至少 1e-9),比这类浮点残留大出几个
    数量级,又远小于任何真实报价的最小变动单位,不会掩盖真实的交叉。

    Tolerance-based float compare, returns -1/0/1.

    Two EMAs built off literally the same repeated prices should be
    identical in theory — but the recursive multiply-add `prev = value*k +
    prev*(1-k)` leaves ~1e-13-to-1e-14-level residue (e.g. 100.0 becomes
    100.00000000000001) whenever the smoothing constant k=2/(period+1) isn't
    exactly representable in binary floating point — which specific periods
    trigger this is unpredictable (period=30 does, 12/26/9/14/50 don't, in
    this codebase's own tests). Feeding that residue straight into a strict
    `<`/`>` comparison against another series misreads it as "just crossed",
    firing a signal out of thin air. The tolerance is 1e-9 relative (1e-9
    floor) — orders of magnitude above that residue, and orders of magnitude
    below any real quote's tick size, so it never masks an actual crossover.
    """
    tol = max(abs(a), abs(b)) * rel_tol + rel_tol
    if a > b + tol:
        return 1
    if a < b - tol:
        return -1
    return 0
